micrometer-core/src/main/java/io/micrometer/core/instrument/observation/ObservationOrTimerCompatibleInstrumentation.java

```diff
@@ -0,0 +1,141 @@
+ * registry, an {@link Observation} will be used for instrumentation. Otherwise, a
+ * {@link Timer} will be used.
+ *
+ * @param <T> context type if Observation used for instrumentation
+ */
+public class ObservationOrTimerCompatibleInstrumentation<T extends Observation.Context> {
+
+    private final MeterRegistry meterRegistry;
+
+    private final ObservationRegistry observationRegistry;
+
+    @Nullable
+    private final Observation.ObservationConvention<T> convention;
+
+    private final Observation.ObservationConvention<T> defaultConvention;
+
+    @Nullable
+    private Timer.Sample sample;
+
+    @Nullable
+    private Observation observation;
+
+    @Nullable
+    private T context;
+
+    /**
+     * Start timing based on Observation and the convention for it if
+     * {@link ObservationRegistry} is not null and not the no-op registry. Otherwise,
+     * timing will be instrumented with a {@link Timer} using the {@link MeterRegistry}.
+     * @param meterRegistry registry for Timer-based instrumentation
+     * @param observationRegistry registry for Observation-based instrumentation
+     * @param context supplier for the context to use if instrumenting with Observation
+     * @param convention convention that overrides the default convention and any
+     * conventions configured on the registry, if not null
+     * @param defaultConvention convention to use if one is not configured
+     * @return a started instrumentation
+     * @param <T> context type if Observation used for instrumentation
+     */
+    public static <T extends Observation.Context> ObservationOrTimerCompatibleInstrumentation<T> start(
+            MeterRegistry meterRegistry, @Nullable ObservationRegistry observationRegistry, Supplier<T> context,
+            @Nullable Observation.ObservationConvention<T> convention,
+            Observation.ObservationConvention<T> defaultConvention) {
+        ObservationOrTimerCompatibleInstrumentation<T> observationOrTimer = new ObservationOrTimerCompatibleInstrumentation<>(
+                meterRegistry, observationRegistry, convention, defaultConvention);
+        observationOrTimer.start(context);
+        return observationOrTimer;
+    }
+
+    private ObservationOrTimerCompatibleInstrumentation(MeterRegistry meterRegistry,
+            @Nullable ObservationRegistry observationRegistry,
+            @Nullable Observation.ObservationConvention<T> convention,
+            Observation.ObservationConvention<T> defaultConvention) {
+        this.meterRegistry = meterRegistry;
+        this.observationRegistry = observationRegistry == null ? ObservationRegistry.NOOP : observationRegistry;
+        this.convention = convention;
+        this.defaultConvention = defaultConvention;
+    }
+
+    private void start(Supplier<T> contextSupplier) {
+        if (observationRegistry.isNoop()) {
+            sample = Timer.start(meterRegistry);
```

Comment:
we have  `sample = ` but then we have `this.context` but then we have `observation = `

Reply:
I wish the format task would normalize this for us automatically. I'll make it consistent.

Reply:
We can have a checkstyle rule for this